# Exploración

Importaciones

In [78]:
import pandas as pd


# 1. Exploración Inicial

In [60]:
df = pd.read_csv('product_activity.csv')
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 8782 entries, 0 to 8781
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   user_id            8782 non-null   str    
 1   created_at         8782 non-null   str    
 2   country            8782 non-null   str    
 3   plan_type          8782 non-null   str    
 4   user_age           8028 non-null   float64
 5   post_id            8782 non-null   str    
 6   post_category      8782 non-null   str    
 7   post_created_at    8782 non-null   str    
 8   votes_received     8782 non-null   int64  
 9   user_total_posts   8782 non-null   int64  
 10  days_since_signup  8782 non-null   int64  
 11  device_type        8782 non-null   str    
dtypes: float64(1), int64(3), str(8)
memory usage: 823.4 KB


In [61]:
df.head()

,user_id,created_at,country,plan_type,user_age,post_id,post_category,post_created_at,votes_received,user_total_posts,days_since_signup,device_type
0,U01988,2025-02-18T02:07:44,PY,pro,26.0,P0008515,sport,2025-05-07T20:55:28,7,16,78,mobile
1,U00236,2025-06-22T07:49:10,BR,free,27.0,P0001023,tech,2025-09-13T20:31:06,1,9,83,web
2,U00791,2024-02-12T02:45:45,CL,free,28.0,P0003405,tech,2024-02-14T05:17:48,11,2,2,mobile
3,U01522,2024-09-22T07:06:50,US,free,16.0,P0006524,finance,2024-09-24T07:51:34,5,2,2,web
4,U01092,2025-07-18T02:27:52,PY,free,NaN,P0004665,education,2025-07-24T04:56:56,7,2,6,mobile


In [62]:
df.describe()

,user_age,votes_received,user_total_posts,days_since_signup
count,8028.000000,8782.000000,8782.000000,8782.000000
mean,27.902591,6.918356,8.324186,29.479390
std,7.547052,5.127311,6.754906,36.819928
min,16.000000,0.000000,1.000000,0.000000
25%,22.000000,3.000000,4.000000,5.000000
50%,28.000000,6.000000,6.000000,17.000000
75%,33.000000,9.000000,11.000000,40.000000
max,58.000000,74.000000,39.000000,404.000000


# Cantidad de datos nulos

In [63]:
# Contamos la cantidad de nulos, se esperan 754

total_null = df.isnull().sum()
print(total_null)

user_id                0
created_at             0
country                0
plan_type              0
user_age             754
post_id                0
post_category          0
post_created_at        0
votes_received         0
user_total_posts       0
days_since_signup      0
device_type            0
dtype: int64


# Cantidad de filas repetidas exactamente

In [64]:
# Contamos la cantidad de filas duplicadas.

duplicated_rows = df.duplicated().sum()
print(duplicated_rows)

172


In [65]:
# Controlamos que la cantidad de post_id duplicados sin que las filas sean exactas.

duplicated_post_id = df.duplicated(subset='post_id').sum()
print(duplicated_post_id)

172


# Cantidad de filas duplicadas: 
Existen 172 filas duplicadas exactas, correspondientes 1:1 con post_id repetidos → limpieza simple, sin ambigüedad

# ¿Cuántos posts ocurren antes del signup?

In [66]:
df['created_at'] = pd.to_datetime(df['created_at'], errors='coerce')
df['post_created_at'] = pd.to_datetime(df['post_created_at'], errors='coerce')
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 8782 entries, 0 to 8781
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   user_id            8782 non-null   str           
 1   created_at         8781 non-null   datetime64[us]
 2   country            8782 non-null   str           
 3   plan_type          8782 non-null   str           
 4   user_age           8028 non-null   float64       
 5   post_id            8782 non-null   str           
 6   post_category      8782 non-null   str           
 7   post_created_at    8781 non-null   datetime64[us]
 8   votes_received     8782 non-null   int64         
 9   user_total_posts   8782 non-null   int64         
 10  days_since_signup  8782 non-null   int64         
 11  device_type        8782 non-null   str           
dtypes: datetime64[us](2), float64(1), int64(3), str(6)
memory usage: 823.4 KB


In [67]:
# controlamos cual es la fila que quedó como NaT razon por la que tenemos solo 8781 en las columnas created_at.

df[df['created_at'].isnull()]


,user_id,created_at,country,plan_type,user_age,post_id,post_category,post_created_at,votes_received,user_total_posts,days_since_signup,device_type
8772,U20001,NaT,PY,pro,29.0,P0099001,tech,2025-03-01 10:00:00,4,3,1,web


1 fecha no parseable en created_at, fila con user_id=U20001, post_id=P0099001 — posible registro corrupto/de prueba

In [68]:
# controlamos cual es la fila que quedó como NaT razon por la que tenemos solo 8781 en las columnas post_created_at.

df[df['post_created_at'].isnull()]

,user_id,created_at,country,plan_type,user_age,post_id,post_category,post_created_at,votes_received,user_total_posts,days_since_signup,device_type
8778,U20007,2025-03-15 12:00:00,MX,enterprise,38.0,P0099007,science,NaT,5,10,10,web


1 fecha no parseable en post_created_at, fila con user_id=U20007, post_id=P0099007 — posible registro corrupto/de prueba

In [69]:
df[df['post_created_at'] < df['created_at']]

,user_id,created_at,country,plan_type,user_age,post_id,post_category,post_created_at,votes_received,user_total_posts,days_since_signup,device_type
22,U01754,2025-01-17 07:05:21,US,free,39.0,P0007502,sports,2024-12-18 16:05:21,3,6,0,web
170,U00707,2024-10-07 17:32:12,MX,pro,27.0,P0003002,gaming,2024-09-27 12:32:12,13,11,0,mobile
204,U00503,2025-01-13 13:27:33,US,FREE,16.0,P0002095,health,2024-12-14 01:27:33,7,9,0,web
208,U01763,2024-12-02 12:42:25,BR,free,39.0,P0007543,music,2024-11-08 11:42:25,5,13,0,web
288,U01122,2024-04-28 09:03:25,US,pro,22.0,P0004783,education,2024-03-29 16:03:25,11,6,0,mobile
...,...,...,...,...,...,...,...,...,...,...,...,...
8281,U00349,2025-12-28 12:28:47,BR,free,26.0,P0001541,tech,2025-12-15 20:28:47,23,1,0,web
8343,U00665,2024-11-03 04:05:20,MX,free,22.0,P0002787,music,2024-10-16 18:05:20,10,1,0,web
8608,U01378,2025-06-19 18:02:45,US,pro,26.0,P0005902,travel,2025-05-20 09:02:45,8,22,0,mobile
8646,U00243,2025-02-03 13:41:27,PY,free,35.0,P0001060,finance,2025-01-31 12:41:27,2,15,0,mobile


In [70]:
posts_antes_signup = df[df['post_created_at'] < df['created_at']]
porcentaje = (len(posts_antes_signup) / len(df)) * 100
print(porcentaje)

1.1386927806877705


# Entonces ¿cuántos post_created_at son anteriores a created_at? 100 - 1.14%

# 2. Limpieza Básica

In [71]:
df['plan_type'].value_counts()

plan_type
free            5978
pro             1460
enterprise       306
Free             208
 free            197
FREE             196
FreE             189
PRo               46
 pro              44
PRO               44
Pro               38
Pro               35
EnterPrise        13
ENTERPRISE        11
 enterprise        7
Enterprise         7
premium            1
vip                1
enterprise+        1
Name: count, dtype: int64

In [72]:
df['device_type'].value_counts()

device_type
web        4416
mobile     3761
desktop     602
tablet        1
console       1
phone         1
Name: count, dtype: int64

In [73]:
df['post_category'].value_counts()

post_category
tech           1187
life            913
sports          899
science         753
finance         739
gaming          727
music           614
health          601
education       592
travel          445
 tech            70
Tech             67
TECH             56
tehc             50
Life             46
Finance          44
 sport           42
sciense          42
gamming          40
 life            38
SPORTS           37
 finance         37
LIFE             36
Sports           36
 science         35
sporst           35
Gaming           35
Science          34
 gaming          33
SCIENCE          33
finanse          32
FINANCE          31
EDUCATION        31
educatoin        31
Music            28
healt            28
 health          26
Education        25
GAMING           22
 education       21
Health           21
MUSIC            21
lfe              21
 music           20
Travel           20
TRAVEL           19
HEALTH           18
 travel          16
trvael           16
musc  

### Normalizamos los valores de las columnas.

In [74]:
plan_type_map = {
    'Free': 'free', 'FREE': 'free', 'FreE': 'free',
    'PRo': 'pro', 'PRO': 'pro', 'Pro': 'pro',
    'EnterPrise': 'enterprise', 'ENTERPRISE': 'enterprise', 
    'Enterprise': 'enterprise','enterprise+': 'enterprise'
}

plan_type_validos = ['free', 'pro', 'enterprise']

df['plan_type'] = df['plan_type'].str.strip()
df['plan_type'] = df['plan_type'].replace(plan_type_map)
df[~df['plan_type'].isin(plan_type_validos)]

,user_id,created_at,country,plan_type,user_age,post_id,post_category,post_created_at,votes_received,user_total_posts,days_since_signup,device_type
8773,U20002,2025-04-10 09:15:00,BR,premium,34.0,P0099002,finance,2025-04-12 12:00:00,8,5,2,mobile
8776,U20005,2025-06-01 00:00:00,CL,vip,27.0,P0099005,mistery,2025-06-02 00:00:00,3,1,1,console


In [81]:
post_category_map = {
    'Tech': 'tech', 'TECH': 'tech', 'tehc': 'tech',
    'Life': 'life', 'LIFE': 'life', 'lfe': 'life',
    'Finance': 'finance', 'finanse': 'finance', 'FINANCE': 'finance',
    'SPORTS': 'sports', 'Sports': 'sports','sporst': 'sports', 'sp0rts': 'sports', 'sport': 'sports',
    'sciense': 'science', 'Science': 'science', 'SCIENCE': 'science',
    'Gaming': 'gaming', 'gamming': 'gaming', 'GAMING': 'gaming',
    'EDUCATION': 'education', 'educatoin': 'education', 'Education': 'education',
    'Music': 'music', 'MUSIC': 'music', 'musc': 'music',
    'healt': 'health', 'Health': 'health', 'HEALTH': 'health',
    'Travel': 'travel', 'TRAVEL': 'travel', 'trvael': 'travel',
}

post_category_validos = ['tech', 'life', 'sports', 'science', 'finance', 'gaming', 'music', 'health', 'education', 'travel']

df['post_category'] = df['post_category'].str.strip()
df['post_category'] = df['post_category'].replace(post_category_map)
df[~df['post_category'].isin(post_category_validos)]

,user_id,created_at,country,plan_type,user_age,post_id,post_category,post_created_at,votes_received,user_total_posts,days_since_signup,device_type
8775,U20004,2024-11-20 14:30:00,AR,pro,41.0,P0099004,politics,2024-11-25 09:00:00,6,7,5,desktop
8776,U20005,2025-06-01 00:00:00,CL,vip,27.0,P0099005,mistery,2025-06-02 00:00:00,3,1,1,console


In [76]:
device_type_map = {
    'phone': 'mobile', 'tablet': 'mobile'
}

device_type_validos = ['web', 'mobile', 'desktop']

df['device_type'] = df['device_type'].str.strip()
df['device_type'] = df['device_type'].replace(device_type_map)
df[~df['device_type'].isin(device_type_validos)]

,user_id,created_at,country,plan_type,user_age,post_id,post_category,post_created_at,votes_received,user_total_posts,days_since_signup,device_type
8776,U20005,2025-06-01,CL,vip,27.0,P0099005,mistery,2025-06-02,3,1,1,console


### Recalculo de fechas

In [ ]:
df['days_since_signup_calc'] = (df['post_created_at'] - df['created_at']).dt.days
df

,user_id,created_at,country,plan_type,user_age,post_id,post_category,post_created_at,votes_received,user_total_posts,days_since_signup,device_type,days_since_signup_calc
0,U01988,2025-02-18 02:07:44,PY,pro,26.0,P0008515,sports,2025-05-07 20:55:28,7,16,78,mobile,78.0
1,U00236,2025-06-22 07:49:10,BR,free,27.0,P0001023,tech,2025-09-13 20:31:06,1,9,83,web,83.0
2,U00791,2024-02-12 02:45:45,CL,free,28.0,P0003405,tech,2024-02-14 05:17:48,11,2,2,mobile,2.0
3,U01522,2024-09-22 07:06:50,US,free,16.0,P0006524,finance,2024-09-24 07:51:34,5,2,2,web,2.0
4,U01092,2025-07-18 02:27:52,PY,free,NaN,P0004665,education,2025-07-24 04:56:56,7,2,6,mobile,6.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
8777,U20006,2025-05-10 00:00:00,PY,free,30.0,P0099006,travel,2025-05-09 23:00:00,2,4,0,web,-1.0
8778,U20007,2025-03-15 12:00:00,MX,enterprise,38.0,P0099007,science,NaT,5,10,10,web,NaN
8779,U20008,2025-07-01 09:00:00,PE,pro,19.0,P0099008,sports,2025-07-02 09:00:00,7,6,1,mobile,1.0
8780,U20009,2025-08-01 10:10:10,CO,enterprise,45.0,P0099009,education,2025-08-03 10:10:10,0,2,2,desktop,2.0


In [85]:
df[df['days_since_signup_calc'] != df['days_since_signup']]

,user_id,created_at,country,plan_type,user_age,post_id,post_category,post_created_at,votes_received,user_total_posts,days_since_signup,device_type,days_since_signup_calc
5,U00488,2025-03-19 16:56:05,US,free,43.0,P0002030,finance,2025-03-28 12:10:25,3,10,9,web,8.0
7,U01863,2024-12-12 21:40:02,CL,free,31.0,P0008000,health,2025-02-02 21:09:32,5,5,52,mobile,51.0
9,U00880,2024-12-12 23:03:33,UY,free,32.0,P0003784,science,2024-12-19 17:52:17,7,2,7,mobile,6.0
17,U01956,2025-02-19 10:47:45,PY,free,16.0,P0008381,finance,2025-04-28 01:34:08,6,11,68,web,67.0
18,U01179,2024-08-28 23:29:24,AR,free,37.0,P0005024,education,2024-09-14 18:31:09,7,4,17,web,16.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
8770,U00196,2025-03-04 20:50:21,US,free,35.0,P0000861,science,2025-03-06 15:03:53,5,6,2,mobile,1.0
8772,U20001,NaT,PY,pro,29.0,P0099001,tech,2025-03-01 10:00:00,4,3,1,web,NaN
8775,U20004,2024-11-20 14:30:00,AR,pro,41.0,P0099004,politics,2024-11-25 09:00:00,6,7,5,desktop,4.0
8777,U20006,2025-05-10 00:00:00,PY,free,30.0,P0099006,travel,2025-05-09 23:00:00,2,4,0,web,-1.0


In [86]:
diferencia = df['days_since_signup_calc'] - df['days_since_signup']
diferencia.describe()

count    8780.000000
mean       -0.702164
std         2.169547
min       -31.000000
25%        -1.000000
50%        -1.000000
75%         0.000000
max         3.000000
dtype: float64